### ISSUE TITLE
Barriers meaning is ambiguous on what should happen to unexpected events.

### ISSUE THREAD

**Issue by dpwspoon** [k3po/k3po#183]
The [barrier documentation](https://github.com/k3po/k3po/wiki/Scripting-Language#barriers) states: 

> Wait for the barrier named “” to be notified before processing upstream events (normally read). Any upstream events that occur before the barrier has been notified will be queued and processed (in order) once the barrier is notified.

and

> if the incoming message arrives before the outgoing message was completely written out, it should be an error. 

These points seem to contradict.  For example

```
read await BARRIER
read "hello, world"
``` 

if "hello, world" arrives before BARRIER has been notified, should it throw an error as stated, or should the event be queued and released once the barrier is complete.

My understanding has been that it should throw an exception.  The code is written such that it would be queued, but the queuing logic has a bug where queued events can race with new events.

**Comment by dpwspoon** [comment]
Ping @jfallows, any input

**Comment by cmebarrow** [comment]
I agree with David, as a test script writer I would expect (and want) the exception. That's kind of the whole point of using a read barrier.

**Comment by dpwspoon** [comment]
Changing this to a bug and re-titling

**Comment by dpwspoon** [comment]
When we fix this we should update the doc
